sdk/src/main/java/com/mapbox/maps/MapController.kt

```diff
@@ -217,7 +218,8 @@ internal class MapController : MapPluginProviderDelegate, MapControllable {
   }
 
   override fun addWidget(widget: Widget) {
-    renderer.renderThread.addWidget(widget)
+    val weakThis = WeakReference(this)
```

Comment:
I'm not sure I understand why we need `WeakReference`.
Isn't it enough to remove it when `removeWidget` is called?

Reply:
User might not call the `removeWidget`, and if the MapView is destroyed(e.g. the fragment that contains the mapview is detached from the view) while user keep the widget instance in their app, there could potentially be a memory leak.

But agree we should clean up the repaintTrigger when widget is removed.

Reply:
Reset the repaint trigger to null when the widget is removed from the map.

Reply:
Discussed offline. `removeWidget` might not be called under all circumstances.

Still, we should add removing the trigger in case the user retains the widget.

Reply:
Could you elaborate please? It seems it's a clear user mistake to keep widgets / (or view annotations) etc created by the MapView

Reply:
And if user does not keep Widgets explicitly - they'll be GC'ed together with the MapView itself, shouldn't be a problem it seems.

Reply:
> It seems it's a clear user mistake to keep widgets / (or view annotations) etc created by the MapView

Actually it's not, the Widget is usually created by user, for example CompassWidget should be created and the instance should be kept by the user, to be able to update the bearing etc later on.

> And if user does not keep Widgets explicitly - they'll be GC'ed together with the MapView itself, shouldn't be a problem it seems.

Not really, user can accidentally keep a reference of the widget even after the MapView is destroyed. Also [this](https://stackoverflow.com/questions/42271208/kotlin-safe-lambdas-no-memory-leak) would be an interesting read.

Reply:
> Actually it's not, the Widget is usually created by user, for example CompassWidget should be created and the instance should be kept by the user, to be able to update the bearing etc later on.

I agree - however how is it different from the View Annotation, for example? User might keep it to update some properties - obviously it will hold the parent MapView as well. It's absolutely the same case IMO.

Reply:
> however how is it different from the View Annotation, for example? User might keep it to update some properties - obviously it will hold the parent MapView as well. It's absolutely the same case IMO.

My argument here is that it's not a user mistake to keep the widgets that they constructed and update it at a later time, actually rather it's an expected usage pattern. I'm not sure I follow the comparison with ViewAnnotation. In this widget's particular case, we are keeping an lamda(potentially keeps a strong MapController) that invokes a method on MapController.

Reply:
@pengdev my 5 cents here are:
1. When using weak ref like now - that indeed potentially saves us from memory leak when map controller is destroyed but widgets (owned by user) are not. However if user will call `widget.triggerRepaint` after controller is destroyed - `weakThis.get()` will be null and this is no-op operation anyway. If same widget will be used with another map controller - in order for that to work widget will need to be re-added to the new map controller.
2. If not using weak ref and use code as now - I could believe we could have the memory leak.

I would actually remove weak ref and explicitly call `widget.setRepaintTrigger(null)` on all widgets that were _not_ removed by the user explicitly in `MapController.onDestroy`. In that case we retain current logic (no leaking) and don't mess with weak refs (we have removed almost all of them from codebase and I do not want to add more of them).

Reply:
> My argument here is that it's not a user mistake to keep the widgets that they constructed and update it at a later time, actually rather it's an expected usage pattern.

Right, but what is expected if user adds widget to the mapView (that owns mapController), then keep it after the mapView/container is destroyed? The parallel with view annotation is that if user creates view annotation, then keeps it while mapView is destroyed - it'll also hold the whole mapView, and this is expected, right?

Reply:
> don't mess with weak refs (we have removed almost all of them from codebase and I do not want to add more of them).

Maybe I'm out of the loop, any reasons we want to avoid the usage of weak refs? just want have a deeper understanding for future reference.

I also agree to clean up all the widgets on MapController.destroy(), and probably that's the best solution here, actually will refactor it that way.

> is expected if user adds widget to the mapView (that owns mapController), then keep it after the mapView/container is destroyed?

It's not expected/preferred, but will be what many users do regardless, in any case, we should avoid a memory leak if possible.

Reply:
> Maybe I'm out of the loop, any reasons we want to avoid the usage of weak refs? just want have a deeper understanding for future reference.

Problem with weakrefs is that they tend to hide the issues with the calling or our code, and then leads to subtle bugs, iirc we've faced them with style/maps/locationengine several times. 

And if we treat widget/child view (view annotation or compass or logo plugin) of the map as a _child_ of the map (this is my understanding of both) its imo it's fine to keep the reference to the parent, if child is stored. It's same case with regular Android classes, if you keep reference to the view - the hosting fragment/activity will also be kept.

Reply:
> Maybe I'm out of the loop, any reasons we want to avoid the usage of weak refs? just want have a deeper understanding for future reference.

I think the most obvious reason (at least that's my IMO) is that that puts our SDK out of control. GC could be called literally at any moment of time by the system and it makes it pretty hard to reproduce / fix any issues connected with that. We had several examples when rare undetermined errors were reported by customers, the latest one I found is https://github.com/mapbox/mapbox-maps-android/pull/1591.

Reply:
Thanks @yunikkk @yunikkk for the thoughts, I'd agree to avoid usage of weak reference if possible, refactored the widgets renderer to do proper clean up during destroy, please take another look.